lib/kube/proxy/self_subject_reviews.go

```diff
@@ -0,0 +1,267 @@
+	resource := depluralizeResource(accessReview.Spec.ResourceAttributes.Resource)
+	name := accessReview.Spec.ResourceAttributes.Name
+	if name == "exec" || name == "attach" || name == "portforward" {
+		name = ""
+	}
+	// If the request is for a resource that Teleport does not support, return
+	// nil and let the Kubernetes API server handle the request.
+	if !slices.Contains(types.KubernetesResourcesKinds, resource) {
+		return nil
+	}
+
+	authPref, err := f.cfg.CachingAuthClient.GetAuthPreference(req.Context())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	state := actx.GetAccessState(authPref)
+	switch err := actx.Checker.CheckAccess(
+		actx.kubeCluster,
+		state,
+		services.RoleMatchers{
+			// Append a matcher that validates if the Kubernetes resource is allowed
+			// by the roles that satisfy the Kubernetes Cluster.
+			&kubernetesResourceMatcher{
+				types.KubernetesResource{
+					Kind:      resource,
+					Name:      name,
+					Namespace: namespace,
+				},
+			},
+		}...); {
+	case errors.Is(err, services.ErrTrustedDeviceRequired):
+		return trace.Wrap(err)
+	case err != nil:
+		accessReview.Status = authv1.SubjectAccessReviewStatus{
+			Allowed: false,
+			Denied:  true,
+			Reason: fmt.Sprintf(
+				"access to %s %s/%s denied by Teleport: please ensure that %q field in your Teleport role defines access to the desired resource.\n\n"+
+					"Valid example:\n"+
+					"kubernetes_resources:\n"+
+					"- kind: %s\n"+
+					"  name: %s\n"+
+					"  namespace: %s\n", resource, namespace, name, kubernetesResourcesKey, resource, emptyOrWildcard(name), emptyOrWildcard(namespace)),
+		}
+
+		responsewriters.SetContentTypeHeader(w, req.Header)
+		if err := encoder.Encode(accessReview, w); err != nil {
+			return trace.Wrap(err)
+		}
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// emptyOrWildcard returns the string s if it is not empty, otherwise it returns
+// '*'.
+func emptyOrWildcard(s string) string {
+	if s == "" {
+		return fmt.Sprintf("'%s'", types.Wildcard)
+	}
+	return s
+}
+
+// parseSelfSubjectAccessReviewRequest parses the request body into a SelfSubjectAccessReview object
+// and replaces the body so it can be read again.
+func parseSelfSubjectAccessReviewRequest(decoder runtime.Decoder, req *http.Request) (*authv1.SelfSubjectAccessReview, error) {
+	payload, err := io.ReadAll(req.Body)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	req.Body.Close()
+
+	req.Body = io.NopCloser(bytes.NewReader(payload))
+	obj, err := decodeAndSetGVK(decoder, payload)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	switch o := obj.(type) {
+	case *authv1.SelfSubjectAccessReview:
+		return o, nil
+	default:
+		return nil, trace.BadParameter("unexpected object type: %T", obj)
+	}
+}
+
+// depluralizeResource returns the singular form of the resource if it is plural.
+func depluralizeResource(resource string) string {
+	if strings.HasSuffix(resource, "s") {
+		return resource[:len(resource)-1]
+	}
+	return resource
+}
```

Comment:
I'm not sure this heuristic is robust enough: `depluralizeResource("ingress")` will have a weird behaviour. 

If you don't think we need to fix this now, can you add a warning in `types.KubernetesResourcesKinds` to ensure we don't introduce support for a resource ending with an `s` without changing the logic here?

Reply:
Right now we only support pods and once we support other options I will have a conversion table